### PR TITLE
feat(COM-235): add the id array as an optional parameter to import as…

### PR DESCRIPTION
**Pull request by mDuval1**
…set method

### CODE REVIEW

File: src/kili/mutations/asset/__init__.py [hunk 1]
```diff
@@ -40,6 +40,7 @@ def append_many_to_dataset(
         project_id: str,
         content_array: Optional[List[str]] = None,
         external_id_array: Optional[List[str]] = None,
+        id_array: Optional[List[str]] = None,
```
Comment (theodu) [inline src/kili/mutations/asset/__init__.py:43]: IdArray is not the ids of the assets so this seems very confusing. Can't you give another name ?

Comment (theodu) [reply src/kili/mutations/asset/__init__.py:43]: and if I understand well, it is for internal use only. @PierreLeveau should be add an underscore before its name ?

Comment (PierreLeveau) [reply src/kili/mutations/asset/__init__.py:43]: like `(for internal use only)`

**Review by PierreLeveau — COMMENTED** [review]
I'm not sure to understand the behavior of this new field:
* what happens if the id is already exisiting? does it replace the content in the asset?
* what happens if the id does not exist yet? is it created ?
* How does it conflict with the external_id_array if they are passed together ?

**Comment by mDuval1** [comment]
@PierreLeveau 
what happens if the id is already exisiting? does it replace the content in the asset?

-> yes, we check the user has proper rights

what happens if the id does not exist yet? is it created ?

-> yes, it's generated. In a second PR, I'll implement the generation of those ids.

How does it conflict with the external_id_array if they are passed together ?

-> It does not. The externalId can be considered a primary key at the project level, the assetId is the primary key of the table

**Comment by mDuval1** [comment]
Reopened, to add assetId generation and filePaths use

File: src/kili/services/asset_import/base.py [hunk 2]
```diff
@@ -177,25 +192,48 @@ def import_batch(self, assets: List[AssetLike]):
...
         return super().import_batch(assets)
 
+    def get_content_type_and_data_from_content(
+        self, content: Union[str, None]
+    ) -> Tuple[bytes, Union[str, None]]:
+        """
+        Returns the data of the content (path) and its content type
+        """
+        assert content
+        with open(content, "rb") as file:
```
Comment (PierreLeveau) [inline src/kili/services/asset_import/base.py:200]: Can you use pathlib? we are migrating away from `os.path`
here:
```suggestion
        with Path(content).open("rb") as file:
```

File: src/kili/services/asset_import/base.py [hunk 3]
```diff
@@ -177,25 +192,48 @@ def import_batch(self, assets: List[AssetLike]):
             assets = self.upload_local_content_to_bucket(assets)
         return super().import_batch(assets)
 
+    def get_content_type_and_data_from_content(
+        self, content: Union[str, None]
```
Comment (PierreLeveau) [inline src/kili/services/asset_import/base.py:194]:
```suggestion
        self, content: Optional[str]
```
here and below

File: src/kili/services/asset_import/base.py [hunk 4]
```diff
@@ -177,25 +192,48 @@ def import_batch(self, assets: List[AssetLike]):
...
+    ) -> Tuple[bytes, Union[str, None]]:
```
Comment (PierreLeveau) [inline src/kili/services/asset_import/base.py:195]:
```suggestion
    ) -> Tuple[bytes, Optional[str]]:
```

File: src/kili/services/asset_import/base.py [hunk 5]
```diff
@@ -177,25 +192,48 @@ def import_batch(self, assets: List[AssetLike]):
...
+        Returns the data of the content (path) and its content type
+        """
+        assert content
+        with open(content, "rb") as file:
+            data = file.read()
+            content_type, _ = mimetypes.guess_type(content)
+            return data, content_type
+
+    def get_type_and_data_from_content_array(
+        self, content_array: List[Union[str, None]]
```
Comment (PierreLeveau) [inline src/kili/services/asset_import/base.py:206]:
```suggestion
        self, content_array: List[Optional[str]]
```

File: src/kili/services/asset_import/base.py [hunk 6]
```diff
@@ -177,25 +192,48 @@ def import_batch(self, assets: List[AssetLike]):
...
+    ) -> List[Tuple[Union[bytes, str], Union[str, None]]]:
```
Comment (PierreLeveau) [inline src/kili/services/asset_import/base.py:207]:
```suggestion
    ) -> List[Tuple[Union[bytes, str], Optional[str]]]:
```

File: src/kili/services/asset_import/base.py [hunk 7]
```diff
@@ -177,25 +192,48 @@ def import_batch(self, assets: List[AssetLike]):
...
-            path = asset.get("content")
-            assert path
-            with open(path, "rb") as file:
-                data_array.append(file.read())
-            content_type, _ = mimetypes.guess_type(path)
-            content_type_array.append(content_type)
+        asset_ids = self.generate_unique_ids(len(assets))
+        project_bucket_path = self.generate_project_bucket_path()
+        asset_content_paths = [
+            os.path.join(project_bucket_path, asset_id, "content") for asset_id in asset_ids
```
Comment (PierreLeveau) [inline src/kili/services/asset_import/base.py:220]:
```suggestion
            Path(project_bucket_path) /  asset_id / "content" for asset_id in asset_ids
```

File: src/kili/services/asset_import/base.py [hunk 8]
```diff
@@ -217,7 +255,12 @@ def upload_json_content_to_bucket(self, assets: List[AssetLike]):
         """
         Upload the json_contents to a bucket with signed urls
         """
-        signed_urls = bucket.request_signed_urls(self.auth, len(assets))
+        asset_ids = self.generate_unique_ids(len(assets))
+        project_bucket_path = self.generate_project_bucket_path()
+        asset_json_content_paths = [
+            os.path.join(project_bucket_path, asset_id, "jsonContent") for asset_id in asset_ids
```
Comment (PierreLeveau) [inline src/kili/services/asset_import/base.py:259]:
```suggestion
            Path(project_bucket_path) / asset_id / "jsonContent" for asset_id in asset_ids
```

File: src/kili/services/asset_import/legacy.py [hunk 9]
```diff
@@ -179,27 +180,38 @@ def upload_content(signed_url: str, content: str, input_type: str):
...
+    if legacy_data_importer.input_type == "TIME_SERIES":
         return list(map(process_time_series, content_array))
     if json_content_array is not None:
         return content_array
     has_local_files = any(not is_url(content) for content in content_array)
     url_content_array = []
     if has_local_files:
-        signed_urls = bucket.request_signed_urls(auth, len(content_array))
+        asset_ids = generate_unique_ids(len(content_array))
+        project_bucket_path = f"projects/{legacy_data_importer.project_id}/assets"
```
Comment (PierreLeveau) [inline src/kili/services/asset_import/legacy.py:205]: here and below, use pathlib

File: src/kili/services/asset_import/video.py [hunk 10]
```diff
@@ -119,7 +121,13 @@ def upload_frames_to_bucket(self, asset: AssetLike):
         """
         frames = asset.get("json_content")
         assert frames
-        signed_urls = bucket.request_signed_urls(self.auth, len(frames))
+        asset_id: str = asset.get("id") or cuid.cuid()
+        project_bucket_path = self.generate_project_bucket_path()
+        asset_frames_paths = [
```
Comment (PierreLeveau) [inline src/kili/services/asset_import/video.py:124]: pathlib

File: src/kili/services/asset_import/legacy.py [hunk 11]
```diff
@@ -179,27 +180,38 @@ def upload_content(signed_url: str, content: str, input_type: str):
...
+    """
+    Returns ids
+    """
+    return [cuid.cuid() for _ in range(size)]
+
+
 # pylint: disable=too-many-arguments
 def process_and_store_content(
-    input_type: str,
+    legacy_data_importer: LegacyDataImporter,
```
Comment (PierreLeveau) [inline src/kili/services/asset_import/legacy.py:190]: passing the full `LegacyDataImporter` is maybe too much (passing two arguments like `input_type` and `Auth`) is ok + it's legacy. @theodu wdyt?

File: src/kili/utils/bucket.py [hunk 12]
```diff
@@ -19,23 +20,27 @@
...
 
 
-def request_signed_urls(auth: KiliAuth, size: int):
+def request_signed_urls(auth: KiliAuth, file_paths: List[str]):
     """
     Get upload signed URLs
     Args:
         auth: Kili Auth
-        project_id: the project Id
         size: the amount of upload signed URL to query
```
Comment (PierreLeveau) [inline src/kili/utils/bucket.py:36]: change docstring

File: src/kili/utils/bucket.py [hunk 13]
```diff
@@ -19,23 +20,27 @@
...
         payload = {
-            "size": nb_url_requested,
+            "filePaths": file_paths,
         }
         urls_response = auth.client.execute(GQL_CREATE_UPLOAD_BUCKET_SIGNED_URLS, payload)
-        signed_urls.extend(urls_response["data"]["urls"])
-    return signed_urls
+        return urls_response["data"]["urls"]
+
+    return functools.reduce(lambda acc, value: acc + value, map(get_file_batch_urls, file_batches))
```
Comment (PierreLeveau) [inline src/kili/utils/bucket.py:51]:
```suggestion
    return [itertools.chain(*map(get_file_batch_urls, file_batches))]
```
?

**Review by PierreLeveau — CHANGES_REQUESTED** [review]

File: src/kili/services/asset_import/base.py (hunk 2 above)
Comment (mDuval1) [reply src/kili/services/asset_import/base.py:200]: done

File: src/kili/services/asset_import/base.py (hunk 3 above)
Comment (mDuval1) [reply src/kili/services/asset_import/base.py:194]: done

File: src/kili/services/asset_import/base.py (hunk 4 above)
Comment (mDuval1) [reply src/kili/services/asset_import/base.py:195]: done

File: src/kili/services/asset_import/base.py (hunk 5 above)
Comment (mDuval1) [reply src/kili/services/asset_import/base.py:206]: done

File: src/kili/services/asset_import/base.py (hunk 6 above)
Comment (mDuval1) [reply src/kili/services/asset_import/base.py:207]: done

File: src/kili/services/asset_import/base.py (hunk 7 above)
Comment (mDuval1) [reply src/kili/services/asset_import/base.py:220]: done

File: src/kili/services/asset_import/base.py (hunk 8 above)
Comment (mDuval1) [reply src/kili/services/asset_import/base.py:259]: done

File: src/kili/services/asset_import/legacy.py (hunk 9 above)
Comment (mDuval1) [reply src/kili/services/asset_import/legacy.py:205]: done

File: src/kili/services/asset_import/video.py (hunk 10 above)
Comment (mDuval1) [reply src/kili/services/asset_import/video.py:124]: done

File: src/kili/services/asset_import/legacy.py (hunk 11 above)
Comment (mDuval1) [reply src/kili/services/asset_import/legacy.py:190]: it's legacy anyway, so I don't want to bother myself with adding multiple parameters

File: src/kili/utils/bucket.py (hunk 12 above)
Comment (mDuval1) [reply src/kili/utils/bucket.py:36]: done

File: src/kili/utils/bucket.py (hunk 13 above)
Comment (mDuval1) [reply src/kili/utils/bucket.py:51]: yes, that's cool didn't know !

**Review by PierreLeveau — APPROVED** [review]

**Review by theodu — COMMENTED** [review]
a

File: tests/test_pagination.py [hunk 14]
```diff
@@ -36,9 +36,9 @@
         "expected_result": ({"id": i} for i in range(20, 220)),
     },
     {
-        "case": "AAU, When I query objects with first=25100, I get the first 25100 objects",
-        "args": {"first": 25100},
-        "expected_result": ({"id": i} for i in range(25100)),
+        "case": "AAU, When I query objects with first=1100, I get the first 1100 objects",
+        "args": {"first": 1100},
+        "expected_result": ({"id": i} for i in range(1100)),
```
[the remark this answers is not in the document]
Comment (mDuval1) [reply tests/test_pagination.py:41]: because it takes close to mn to run.

**Review by theodu — APPROVED** [review]

**Pull request merged**